### PR TITLE
LF: Simplify archive reader.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
- decouple `Reader` and `Decoder`
- introduce case class to handle hash, proto payload, and version
- remove the `..AndVersion` methods for Reader/Decoder that was not very useful. 

as a result this simplifies both the code of the reader and decoders **and** the usage of them.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
cc @oliverse-da 
Those changes may be breaking for Canton. Hopefully it will allow to simplify canton code as well.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Reader.scala [hunk 1]
```diff
@@ -130,19 +106,13 @@ object Reader extends Reader[(PackageId, DamlLf.ArchivePayload)] {
...
-  // Archive Reader that just checks package hash.
-  val HashChecker = new Reader[Unit] {
-    override protected[this] def readArchivePayloadOfVersion(
-        hash: PackageId,
-        lf: DamlLf.ArchivePayload,
-        version: LanguageVersion,
-    ): Unit = ()
-  }
-
+  def apply(recursionLimit: Int = PROTOBUF_RECURSION_LIMIT): Reader = new Reader(recursionLimit)
```
Comment (cocreature) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Reader.scala:109]: at that point maybe just make it a case class? it’s immutable anyway afaict

File: daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala [hunk 2]
```diff
@@ -18,24 +18,21 @@ object LoadDarFunction extends App {
 
   def load(darFile: File, base: String, funcName: String): (Long => Long) = {
 
-    val packages = UniversalArchiveReader().readFile(darFile).get
-    val packagesMap =
-      packages.all.map { case (pkgId, pkgArchive) =>
-        Decode.readArchivePayloadAndVersion(pkgId, pkgArchive)._1
-      }.toMap
+    val paylaod = UniversalArchiveReader().readFile(darFile).get
+    val packages = paylaod.all.map(Decode.decode).toMap
```
Comment (cocreature) [inline daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala:22]:
```suggestion
    val payload = UniversalArchiveReader().readFile(darFile).get
    val packages = payoad.all.map(Decode.decode).toMap
```

File: daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala [hunk 3]
```diff
@@ -18,24 +18,21 @@ object LoadDarFunction extends App {
...
 
     val compiledPackages: CompiledPackages =
-      PureCompiledPackages.assertBuild(packagesMap, compilerConfig)
+      PureCompiledPackages.assertBuild(packages, compilerConfig)
 
     def function(argValue: Long): Long = {
       val expr: SExpr = {
         val ref: DefinitionRef =
-          Identifier(packages.main._1, QualifiedName.assertFromString(s"${base}:${funcName}"))
+          Identifier(paylaod.main.pkgId, QualifiedName.assertFromString(s"${base}:${funcName}"))
```
Comment (cocreature) [inline daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala:35]:
```suggestion
          Identifier(payload.main.pkgId, QualifiedName.assertFromString(s"${base}:${funcName}"))
```

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Reader.scala [hunk 4]
```diff
@@ -12,70 +12,48 @@ import com.daml.lf.language.{LanguageMajorVersion, LanguageVersion}
 import com.daml.daml_lf_dev.DamlLf
 import com.google.protobuf.CodedInputStream
 
-abstract class Reader[+Pkg] {
+final class Reader(recursionLimit: Int) {
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Reader.scala]: change `class` to `object`

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Reader.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/Reader.scala]: dropped.

File: daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala]: fixed. Thanks.

File: daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/perf/src/main/scala/com/daml/lf/explore/LoadDarFunction.scala]: fixed.

**Pull request merged**